lxd/instance/drivers/driver_lxc.go

```diff
@@ -6032,12 +6069,12 @@ func (d *lxc) insertMountLXC(source, target, fstype string, flags int) error {
 	return nil
 }
 
-func (d *lxc) insertMount(source, target, fstype string, flags int, shiftfs bool) error {
-	if d.state.OS.LXCFeatures["mount_injection_file"] && !shiftfs {
+func (d *lxc) insertMount(source, target, fstype string, flags int, idmapped string) error {
```

Comment:
@brauner I do think it might be beneficial to define the `idmapped` arg here as its own type, and then also define string constants for the supported values, if for no other reason then to provide a form of documentation as to what can be passed into this function. It can also help with finding references to those constants in the future, that is more reliable than, for example, searching the code base for "none".

Also, one thing I don't follow here is the difference between the "none" and empty string "" values for idmapped, the two seem to be equivalent from a language perspective, but looking at this function they are treated differently.

Reply:
> @brauner I do think it might be beneficial to define the `idmapped` arg here as its own type, and then also define string constants for the supported values, if for no other reason then to provide a form of documentation as to what can be passed into this function. It can also help with finding references to those constants in the future, that is more reliable than, for example, searching the code base for "none", as we can use features built into the editor to find uses of those constants.

Ok, makes sense.

> 
> Also, one thing I don't follow here is the difference between the "none" and empty string "" values for idmapped, the two 

"none" is easier to understand. It better conveys the information that no idmapping is used than "". But this argument only holds if we open-code and don't introduce a new type. With the new type we can encode this information in the enum names directly.

> seem to be equivalent from a language (English) perspective, but looking at this function they are treated differently.

Hm, I'm not sure how they are treated differently.

Reply:
One of the reasons for not using string constants was that we need them available in `lxd/seccomp/seccomp.go` and I think there's some include/package nonsensery there.

Reply:
> One of the reasons for not using string constants was that we need them available in `lxd/seccomp/seccomp.go` and I think there's some include/package nonsensery there.

Ah yeah good point, "github.com/lxc/lxd/lxd/seccomp" is imported by the lxc instance driver, so the `seccomp` package cannot import the instance otherwise it would create an import loop.

You could move the type and constants to a new tiny package called something like `lxd/instance/idmap` similar to `lxd/instance/instancetype` which contains shared types which had circular import issues previously (this new package might also be a useful place to put file shifting functions too later on too). 

And then the function definition would become something like:

```go
insertMount(source, target, fstype string, flags int, idmapMode idmap.Mode) 
insertMount(..., idmap.ModeShiftfs)
```

But its not a hill I'll die on, so will leave it up to you. :)

Reply:
I just moved them into the idmap package itself as that works fine and it's a good enough location. :)

Reply:
@brauner oh nice, I'd forgotten about that package, that's ideal. You may find some of the automated go lint tests complain about "stuttering" for using `idmap.Idmap...` names though, so `idmap.StorageType...` might appease them. :)

Reply:
Ah, I was thinking about this locally but `golint` didn't complain. What package do I need to install to see these complaints? :)

Reply:
@brauner I ran `golint ./...` in the LXD source tree and it lists this as a relevant existing example:

```
shared/idmap/idmapset_linux.go:268:6: type name will be used as idmap.IdmapSet by other packages, and that stutters; consider calling this Set
```